### DISCUSSION TITLE
[Feature Request] Foundry Forge Tests

### DISCUSSION THREAD

**Discussion by gtsui** [zkSync-Community-Hub/zksync-developers#70] [Feature requests]
### Network

zkSync Era

### Provide a brief description of the feature you would like to see implemented.

Currently several features from foundry are compatible for zksync. However, one of the most crucial tools that I personally, and I'm sure a lot of other devs, use is foundry forge tests forking mainnet. I understand this feature may not be easily implementable, but are there plans to support this any time in the near future? Thanks!

### Who does this feature request benefit the most (users, developers, protocol, etc.)?

developers, protocols

**Comment by uF4No** [comment]
@gtsui yes! To support forking on tests we'd need to integrate the [just released in-memory node](https://era.zksync.io/docs/tools/testing/era-test-node.html) with our [custom Foundry implementation](https://github.com/matter-labs/foundry-zksync). In addition, we'll also integrate it with Hardhat by the way.

If you need to run simulate transactions against a mainnet fork now, [you can fork mainnet following the instructions here](https://era.zksync.io/docs/tools/testing/era-test-node.html#forking-networks) and run scripts agains the in-memory node, which runs in `http://localhost:8011`.

Note that the in-memory node is still in alpha so, if you find any issues you can report them [in its repository](https://github.com/matter-labs/era-test-node)

**Comment by sjewisclare** [comment]
Implementing the feature of forking mainnet for testing purposes, similar to Foundry Forge Tests, would greatly benefit developers and protocol development on zkSync Era. However, it may require significant development effort and infrastructure support, so it's essential to assess its feasibility and prioritize accordingly.

**Comment by tysaboswe** [comment]
Implementing support for forking the mainnet for testing purposes in zkSync Era would greatly benefit developers who rely on this feature for testing and development of their applications and protocols. This functionality would enhance the developer experience and facilitate the creation of robust and reliable applications on the zkSync platform.

**Comment by sinanceviccarter** [comment]
Feature Request: Implement foundry forge tests for forking mainnet on zkSync Era.
Beneficiaries: Developers and protocols.

**Comment by timmeriweather** [comment]
Supporting foundry forge tests for forking mainnet on zkSync Era would greatly benefit developers and protocols by enabling more efficient testing and development workflows.

**Comment by estacilo** [comment]
As of now, there are no official plans to support forking the zkSync mainnet for testing purposes using Foundry Forge. However, the zkSync team is constantly working on improving the development experience and may consider adding this feature in the future. In the meantime, developers can use the zksync-local package to create a local development environment for zkSync and test their applications in a local environment that mimics the mainnet.

**Comment by dutterbutter** [comment]
This is completed.
